File: rubykaigi2026-codeparty/sample1.c
#include <stdio.h>
#include <stdlib.h>

struct Data
{
    int value;
};

struct Data *create_value()
{
    struct Data x = {42};
    return &x; // ローカル変数のアドレスを返している！
}

int main()
{
    struct Data *p = create_value();
    // pはすでに無効なメモリを指している（dangling pointer）
    printf("%d\n", p->value); // 未定義動作！
    return 0;
}

File: rubykaigi2026-codeparty/rust-sample/examples/sample1.rs
struct Data {
    value: i32,
}

fn main() {
    let p: &Data = create_value();
    //     ^^^^
    // pはmain()のスコープが終わるまで有効な参照を期待している
    // → ライフタイム 'a はmain()のスコープと同じ長さが必要

    println!("{}", p.value); // ← ここでまだ使いたい！
}

fn create_value<'a>() -> &'a Data {
    let x = Data { value: 42 };
    &x // コンパイルエラー
}
